[installer] Add EKS installer test

nandajavarma opened the pull request:
## Description
<!-- Describe your changes in detail -->
This PR adds the `EKS` terraform module for testing purposes to the  repo. These are the most important updates:
- The [EKS terraform module](https://github.com/gitpod-io/gitpod/pull/10709/files#diff-8f2a6882e07a098789798ca13ae6a2c42aa710493839d630a73a729b5409cbd3) added to `infra` directory - the security groups are at the moment very permissive. This can be fixed as a part of the epic to make these tf module production ready if need be.
-  changes to [`installer-test.ts`](https://github.com/gitpod-io/gitpod/pull/10709/files#diff-073c7685dea478fdf0c74dd43a01d49232bd3b941d8dac00658e9a885bbfb1e8R35), `Makefile` and `main.tf` to accommodate new `EKS` test configuration
-  [`eks-installer-test.yml` ](https://github.com/gitpod-io/gitpod/pull/10709/files#diff-a767948566688b035be3ec5880960f24e8de5bd3fe109084c22a3565f09fdd0cR1) to trigger the cronjob for this configuration
- standardizing the `external-dns` and `issuer` modules to be use used for both `azure` and `aws`. This involves:
   - [changing the `external_dns_settings` output from aks module](https://github.com/gitpod-io/gitpod/pull/10709/files#diff-9dfe0ce3882a86f9870fa7c9a20c2a0f35d0712d7ba86876f189ced6018a4cedR39) to support `dynamic` setting in external-dns tf module.
   -  renaming module `external-dns` to `cloud-dns-external-dns`(since this is only for cloud-dns - this will be cleaned up more in a future PR), renaming `azure-external-dns` to `external-dns` since this will eventually be standard across all providers
- Added extra [KOTS configs to support aws dependencies](https://github.com/gitpod-io/gitpod/pull/10709/files#diff-8b980a1d07ae91552e229e45e8a4bf39fc10797922d64f82e9d08aacbbb36659) (storage, registry, db)
- Updating the [output.tf in the test directory ](https://github.com/gitpod-io/gitpod/pull/10709/files#diff-a08f9a5da9bf9c8bb7de2a77401d017e5ced2cdb35d9079fa77d3fdfe8cdaecb)to optionally get aws or azure outputs (this will be expanding to GCP) in a later PR 

## Related Issue(s)
<!-- List the issue(s) this PR solves -->
Fixes #

## How to test
<!-- Provide steps to test this PR -->
To test this, get authentication to `core-dev` cluster on GCP. Then run:

```
werft run github -f -s .werft/installer-tests.ts -j .werft/eks-installer-tests.yaml -a debug=true
```

## Release Notes
<!--
  Add entries for the CHANGELOG.md or "NONE" if there aren't any user facing changes.
  Each line becomes a separate entry.
  Format: [!<optional for breaking>] <description>
  Example: !basic auth is no longer supported
  See https://www.notion.so/gitpod/Release-Notes-513a74fdd23b4cb1b3b3aefb1d34a3e0
-->
```release-note
NONE
```

## Documentation
<!--
Does this PR require updates to the documentation at www.gitpod.io/docs?
* Yes
  * 1. Please create a docs issue: https://github.com/gitpod-io/website/issues/new?labels=documentation&template=DOCS-NEW-FEATURE.yml&title=%5BDocs+-+New+Feature%5D%3A+%3Cyour+feature+name+here%3E
  * 2. Paste the link to the docs issue below this comment
* No
  * Are you sure? If so, nothing to do here.
-->

roboquat commented:
@nandajavarma: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

adrienthebo reviewed (approved):
Well, blast; I had a long, throughful approval message and GitHub ate it. Here's the short version!

This is really great work, and I see the effort that you put in! This'll be really helpful for improving our developer velocity and solidifying the self hosted releases.

In short, we can ship this as is. I see some risks, outlined below, but I'm approving this and will leave any changes up to your discretion, @nandajavarma.

- - -

The following items can be optionally extracted to separate issues. Since I'm asking for the work I can pick up the issue creation and contribute fixes; let me know if you'd like support here.

**Blockers**

None!

**Risks/recommendations**

- [ ] Hardcoded database passwords: this is the kind of thing that could bite us down the road. We can hopefully drop in a terraform `random_password` resource to solve this. Would prefer if we do that before merging, can address that afterwards.
- [ ] VPC ingress/egress rules: our customers are going to have more restrictive VPC rules; we'll need to prove out what's actually needed for everything to work.

**Suggestions**

- [x] Factor our `terraform apply -target`: we're leaning on this feature pretty heavily and it's meant to be a niche behavior. We're likely to stretch this a little too far in the foreseeable future. We should _definitely_ handle this one later but let's keep an eye on removing the need for targeted applies.

**Small fixups**

I'm still a walking spell checker 😅 Any fixups for spelling or small details are opportunistic fixes and don't impact correctness or behavior. Adjust as you see fit but don't sweat them!

- - -

Again, thanks for all of the hard work!

nandajavarma commented on install/infra/terraform/eks/dns.tf:
```diff
@@ -0,0 +1,114 @@
+  name = var.domain_name
+
+  tags = {
+    Environment = "test"
```
Ah no, as of now we do not depend on anything with these. I was thinking maybe it will be useful to cleanup orphan resources. But nothing is implemented.

nandajavarma commented on install/infra/terraform/eks/variables.tf:
```diff
@@ -0,0 +1,70 @@
+variable "cluster_name" {
+  type        = string
+  description = "EKS cluster name."
+  default = "nan-eks-3"
```
So we use the randomly generated `TEST_ID` as cluster_name in the testing `main.tf` already. I will drop the name `nan-eks-3`, makes no sense XD

nandajavarma commented on install/tests/Makefile:
```diff
@@ -29,54 +29,68 @@ gke-standard-cluster:
+	terraform apply -target=module.$(cloud)-add-dns-record  -var kubeconfig=${KUBECONFIG} --auto-approve
+	@echo "Done adding NS record"
 
+cloud ?= azure
```
This vairable is expected to set as an argument to the make target, something like `make cluster-issuer cloud=aws` to choose the appropriate module. I am replacing this behavior with an environment variable[ in a follow up PR.](https://github.com/gitpod-io/gitpod/pull/11040/files)

adrienthebo replied:
Cool, thanks for clarifying!

nandajavarma commented on install/infra/terraform/eks/kubernetes.tf:
```diff
@@ -0,0 +1,142 @@
+  name = "nodes-sg-${var.cluster_name}"
+  vpc_id = module.vpc.vpc_id
+
+  ingress {
```
made an issue in a follow-up epic https://github.com/gitpod-io/gitpod/issues/11027

nandajavarma commented on .werft/installer-tests.ts:
```diff
@@ -254,7 +281,10 @@ function cleanup() {
 
-    const response = exec(`make -C ${makefilePath} cleanup`, { slice: "run-terrafrom-destroy", dontCheckRc: true });
+    const response = exec(`make -C ${makefilePath} cleanup cloud=${cloud}`, {
+        slice: "run-terrafrom-destroy",
```
correct!

nandajavarma commented on install/infra/terraform/eks/database.tf:
```diff
@@ -0,0 +1,40 @@
+  name = "dh-sg-${var.cluster_name}"
+  vpc_id = module.vpc.vpc_id
+
+  ingress {
```
:heavy_check_mark:

nandajavarma commented:
Updates made:
- Removed unnecessary secret mount on `eks-installer-tests.yaml`
- Fixed a couple of typos
- Using random password for database
- set `var-region` insterad of hardcoded region in output
- increased ttl for dns zone
- dropped harcoded default `cluster_name` for EKS

nandajavarma commented on .werft/eks-installer-tests.yaml:
```diff
@@ -0,0 +1,88 @@
+  - name: sh-playground-dns-perm
+    secret:
+      secretName: sh-playground-dns-perm
+  - name: sh-aks-perm
```
done!

nandajavarma commented:
/unhold

Letting this one in since all the review suggestions are acknowledged.
